src/gui/dialogs/units_dialog.hpp

```diff
@@ -0,0 +1,240 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include "gui/dialogs/modal_dialog.hpp"
+#include "gui/widgets/button.hpp"
+#include "gui/widgets/group.hpp"
+#include "gui/widgets/listbox.hpp"
+#include "gui/widgets/toggle_button.hpp"
+#include "gui/widgets/unit_preview_pane.hpp"
+#include "team.hpp"
+#include "units/ptr.hpp"
+#include "units/types.hpp"
+#include "utils/optional_fwd.hpp"
+
+#include <boost/dynamic_bitset/dynamic_bitset.hpp>
+#include <string>
+#include <vector>
+
+class team;
+
+namespace gui2::dialogs
+{
+
+class units_dialog : public modal_dialog
+{
+public:
+	units_dialog();
+
+	int get_selected_index() const
+	{
+		return selected_index_;
+	}
+
+	bool is_selected() const
+	{
+		return selected_index_ != -1;
+	}
+
+	utils::optional<unit_const_ptr> get_unit() const
+	{
+		if(is_selected() || !recall_list_.empty()) {
+			return recall_list_[get_selected_index()];
+		} else {
+			return {};
+		}
+	}
+
+	utils::optional<const unit_type*> get_type() const
+	{
+		if(is_selected() || !recruit_list_.empty()) {
+			return recruit_list_[get_selected_index()];
+		} else {
+			return {};
+		}
+	}
+
+	/** Gender choice from the user. */
+	unit_race::GENDER gender() const
+	{
+		return gender_;
+	}
+
+	/** Variation choice from the user. */
+	std::string variation() const
+	{
+		return variation_;
+	}
+
+	units_dialog& set_title(const std::string& title)
+	{
+		title_ = title;
+		return *this;
+	}
+
+	units_dialog& set_ok_label(const std::string& ok_label)
+	{
+		ok_label_ = ok_label;
+		return *this;
+	}
+
+	units_dialog& show_variations(const bool show_variation)
+	{
+		show_variation_grid_ = show_variation;
+		return *this;
+	}
+
+	units_dialog& show_gender(const bool show_gender)
+	{
+		show_gender_grid_ = show_gender;
+		return *this;
+	}
+
+	units_dialog& show_dismiss_option(const bool show_dismiss)
+	{
+		show_dismiss_ = show_dismiss;
+		return *this;
+	}
+
+	units_dialog& show_rename_option(const bool show_rename)
+	{
+		show_rename_ = show_rename;
+		return *this;
+	}
+
+	units_dialog& show_all_headers()
+	{
+		visible_headers_.set();
+		return *this;
+	}
+
+	units_dialog& hide_all_headers()
+	{
+		visible_headers_.reset();
+		return *this;
+	}
+
+	units_dialog& show_header(const int col_num, const bool visible = true)
+	{
+		visible_headers_[col_num] = visible;
+		return *this;
+	}
+
+	units_dialog& set_help_topic(const std::string& topic_id)
+	{
+		topic_id_ = topic_id;
+		return *this;
+	}
+
+	units_dialog& set_team(team* team)
+	{
+		team_ = team;
+		return *this;
+	}
+
+	units_dialog& set_types(std::vector<const unit_type*> types)
+	{
+		recruit_list_ = types;
+		return *this;
+	}
+
+	units_dialog& set_units(std::vector<unit_const_ptr> units)
```

Comment:
Can this take an rvalue-reference? Either way, should assign recall_list with std::move.

Reply:
could probably be dropped, i don't think i'm using team anywhere, see the builders in menu_events.cpp, team is coming via ref capture in lambdas.

Reply:
Even better!

Reply:
apparent team is being used so set_team can't be dropped, but changed to rvalue refs and std::move.

Reply:
Please check, seems like I have to use `std::move` twice now, seems a bit suspicious.

Reply:
also renamed to `recruit_list_` and `recall_list_` to `unit_type_list_` and `unit_list_`.